### PR TITLE
Added `http_method`, `headers` and `ca_file_path` fields to `ExternalIntegration` model.

### PR DESCRIPTION
**Pull request by HanlinMiao**
# Closes: #4895 
# What's Changed

1. Added http_method, headers, and ca_file_path fields to ExternalIntegration model. 
2. Added accordion toggle to ExternalIntegrationEdit/CreateForm 
3. Added helper method to render remote_url, headers, and extra_config in jinja2 templates.

https://github.com/nautobot/nautobot/assets/46973263/d9796892-1944-4436-ace7-1373c9119343

# TODO

- [x] Explanation of Change(s)
- [x] Added change log fragment(s) (for more information see [the documentation](https://docs.nautobot.com/projects/core/en/stable/development/#creating-changelog-fragments))
- [x] Attached Screenshots, Payload Example
- [ ] Unit, Integration Tests
- [ ] Documentation Updates (when adding/changing features)
- [ ] Example Plugin Updates (when adding/changing features)
- [ ] Outline Remaining Work, Constraints from Design

### CODE REVIEW

File: nautobot/extras/migrations/0103_add_extra_fields_to_external_integration_model.py
Comment (glennmatthews) [inline nautobot/extras/migrations/0103_add_extra_fields_to_external_integration_model.py:1]: Would it be preferable to make this an addition to the existing `0101` migration, rather than adding a new one, since we haven't shipped a final release with that migration in it yet?

File: nautobot/extras/forms/forms.py [hunk 1]
```diff
@@ -605,10 +609,17 @@ class ExternalIntegrationBulkEditForm(NautobotBulkEditForm):
     verify_ssl = forms.NullBooleanField(required=False, label="Verify SSL", widget=BulkEditNullBooleanSelect)
     timeout = forms.IntegerField(required=False, min_value=0)
     extra_config = forms.JSONField(required=False)
+    http_method = forms.ChoiceField(
+        required=False,
+        label="HTTP Method",
+        choices=add_blank_choice(HTTPMethodClassChoices),
+    )
+    headers = forms.JSONField(required=False, label="HTTP Request headers")
+    ca_file_path = forms.CharField(required=False, label="CA File Path")
```
Comment (glennmatthews) [inline nautobot/extras/forms/forms.py:618]: Is this order of fields in the form the best/most logical way to structure them? I would have thought `extra_config` at least should be the last in the form.

File: nautobot/extras/forms/forms.py [hunk 2]
```diff
@@ -605,10 +609,17 @@ class ExternalIntegrationBulkEditForm(NautobotBulkEditForm):
...
 
     class Meta:
         model = ExternalIntegration
-        nullable_fields = ["extra_config", "secrets_group"]
+        nullable_fields = ["extra_config", "secrets_group", "http_method", "headers", "ca_file_path"]
```
Comment (glennmatthews) [inline nautobot/extras/forms/forms.py:622]: Looking at the model, `http_method` and `ca_file_path` are not nullable, nor should they be.

File: nautobot/extras/choices.py [hunk 3]
```diff
@@ -125,6 +125,27 @@ class DynamicGroupOperatorChoices(ChoiceSet):
     )
 
 
+#
+# HTTP Methods
```
Comment (glennmatthews) [inline nautobot/extras/choices.py:129]: Why not re-use the existing WebhookHttpMethodChoices instead of duplicating it?

File: nautobot/extras/models/models.py [hunk 4]
```diff
@@ -489,13 +490,55 @@ class ExternalIntegration(PrimaryModel):
         null=True,
         help_text="Optional user-defined JSON data for this integration",
     )
+    http_method = models.CharField(
+        max_length=10,
+        choices=HTTPMethodClassChoices,
+        default=HTTPMethodClassChoices.METHOD_GET,
```
Comment (glennmatthews) [inline nautobot/extras/models/models.py:496]: Given that it's blankable, does it need a default other than ""?

File: nautobot/extras/models/models.py [hunk 5]
```diff
@@ -489,13 +490,55 @@ class ExternalIntegration(PrimaryModel):
...
+        blank=True,
+    )
+    headers = models.JSONField(
+        encoder=DjangoJSONEncoder,
+        blank=True,
+        null=True,
+        help_text="Headers for the HTTP request",
+    )
+    ca_file_path = models.CharField(
+        max_length=100,
```
Comment (glennmatthews) [inline nautobot/extras/models/models.py:507]: Any reason for this limit or can we use something larger like 255?

File: nautobot/extras/models/models.py [hunk 6]
```diff
@@ -489,13 +490,55 @@ class ExternalIntegration(PrimaryModel):
...
+        """
+        Render extra_config and return a dict of Header: Value pairs.
+        """
+        if not self.extra_config:
+            return {}
+        data = render_jinja2(self.extra_config, context)
+        return data
+
+    def render_remote_url(self, context):
+        if not self.remote_url:
```
Comment (glennmatthews) [inline nautobot/extras/models/models.py:537]: remote_url should never be blank

File: nautobot/extras/tables.py [hunk 7]
```diff
@@ -423,6 +425,8 @@ class Meta(BaseTable.Meta):
             "secrets_group",
             "verify_ssl",
             "timeout",
+            "http_method",
+            "ca_file_path",
         )
```
Comment (glennmatthews) [inline nautobot/extras/tables.py:430]: Looks like we're missing `tags` as a column, would you mind adding it?

File: nautobot/extras/models/models.py [hunk 8]
```diff
@@ -489,13 +490,55 @@ class ExternalIntegration(PrimaryModel):
...
+    )
+    headers = models.JSONField(
+        encoder=DjangoJSONEncoder,
+        blank=True,
+        null=True,
+        help_text="Headers for the HTTP request",
+    )
+    ca_file_path = models.CharField(
+        max_length=100,
+        blank=True,
```
Comment (gsnider2195) [inline nautobot/extras/models/models.py:508]: Can you add `verbose_name = "CA file path",`?

File: nautobot/extras/models/models.py [hunk 9]
```diff
@@ -489,13 +490,55 @@ class ExternalIntegration(PrimaryModel):
...
+        data = render_jinja2(self.headers, context)
+        return data
+
+    def render_extra_config(self, context):
+        """
+        Render extra_config and return a dict of Header: Value pairs.
+        """
+        if not self.extra_config:
+            return {}
+        data = render_jinja2(self.extra_config, context)
```
Comment (gsnider2195) [inline nautobot/extras/models/models.py:533]: `render_jinja2` returns a string so I don't think we should return an empty dict if extra_config is blank

File: nautobot/extras/models/models.py [hunk 10]
```diff
@@ -489,13 +490,55 @@ class ExternalIntegration(PrimaryModel):
...
     class Meta:
         ordering = ["name"]
 
+    def render_headers(self, context):
+        """
+        Render headers and return a dict of Header: Value pairs.
+        """
+        if not self.headers:
+            return {}
+        data = render_jinja2(self.headers, context)
```
Comment (gsnider2195) [inline nautobot/extras/models/models.py:524]: Same comment as below

File: nautobot/extras/templates/extras/externalintegration_update.html [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+        <div class="panel-heading"><strong>External Integration</strong></div>
+        <div class="panel-body">
+            {% render_field form.name %}
+            {% render_field form.remote_url %}
+            {% render_field form.secrets_group %}
+            {% render_field form.timeout %}
+            {% render_field form.extra_config %}
+        </div>
+    </div>
+    <div id="accordion" class="panel panel-default">
```
[the remark this answers is not in the document]
Comment (HanlinMiao) [reply nautobot/extras/templates/extras/externalintegration_update.html:16]: What do you think @gsnider2195?

**Comment by HanlinMiao** [comment]
Note to self, need to add those fields to ExternalIntegrationFactory as well.

File: nautobot/extras/models/models.py [hunk 12]
```diff
@@ -489,13 +489,53 @@ class ExternalIntegration(PrimaryModel):
...
         ordering = ["name"]
 
+    def render_headers(self, context):
+        """
+        Render headers and return a dict of Header: Value pairs.
+        """
+        if not self.headers:
+            return ""
+        data = render_jinja2(self.headers, context)
+        return data
```
Comment (glennmatthews) [inline nautobot/extras/models/models.py:524]: Docstring/behavior mismatch here. Also given `self.headers` is a JSON field, do we need to stringify it before rendering as jinja2? Perhaps something like:

```suggestion
        Render headers and return a dict of Header: Value pairs.
        """
        if not self.headers:
            return {}
        data = json.loads(render_jinja2(json.dumps(self.headers, ensure_ascii=False), context)
        return data
```

Probably need to also add error handling in case render_jinja2 or json.loads throws an exception.

File: nautobot/extras/models/models.py [hunk 13]
```diff
@@ -489,13 +489,53 @@ class ExternalIntegration(PrimaryModel):
...
+
+    def render_extra_config(self, context):
+        """
+        Render extra_config and return a dict of Header: Value pairs.
+        """
+        if not self.extra_config:
+            return ""
+        data = render_jinja2(self.extra_config, context)
+        return data
```
Comment (glennmatthews) [inline nautobot/extras/models/models.py:533]: Same comments here.

File: nautobot/extras/models/models.py [hunk 14]
```diff
@@ -489,13 +489,53 @@ class ExternalIntegration(PrimaryModel):
...
+
+    def render_remote_url(self, context):
```
Comment (glennmatthews) [inline nautobot/extras/models/models.py:535]: Missing docstring?

File: nautobot/extras/models/models.py [hunk 15]
```diff
@@ -489,13 +489,53 @@ class ExternalIntegration(PrimaryModel):
...
+        verbose_name="CA file path",
+    )
 
     def __str__(self):
         return f"{self.name} ({self.remote_url})"
 
     class Meta:
         ordering = ["name"]
 
+    def render_headers(self, context):
```
Comment (glennmatthews) [inline nautobot/extras/models/models.py:517]: Need some model tests for these `render_*` APIs including negative tests.

File: nautobot/extras/tables.py [hunk 16]
```diff
@@ -431,6 +437,8 @@ class Meta(BaseTable.Meta):
             "secrets_group",
             "verify_ssl",
             "timeout",
+            "http_method",
```
Comment (glennmatthews) [inline nautobot/extras/tables.py:440]: should this be moved up?

File: changes/4895.added [hunk 17]
```diff
@@ -0,0 +1 @@
+Added `http_method`, `headers` and `ca_file_path` fields to `ExternalIntegration` model.
```
Comment (glennmatthews) [inline changes/4895.added:1]:
```suggestion
Added `http_method`, `headers`, and `ca_file_path` fields, as well as `render_headers`, `render_extra_config`, and `render_remote_url` APIs, to the `ExternalIntegration` model.
```

File: nautobot/extras/models/models.py [hunk 18]
```diff
@@ -489,13 +489,66 @@ class ExternalIntegration(PrimaryModel):
...
+        """
+        Render headers and return a dict of Header: Value pairs.
+        Raises:
```
Comment (glennmatthews) [inline nautobot/extras/models/models.py:520]: Docstring convention is to have a blank line after the first line of the docstring.

```suggestion
        Render headers and return a dict of Header: Value pairs.

        Raises:
```

File: nautobot/extras/models/models.py [hunk 19]
```diff
@@ -489,13 +489,66 @@ class ExternalIntegration(PrimaryModel):
...
+        if not self.headers:
+            return {}
+
+        data = json.loads(render_jinja2(json.dumps(self.headers, ensure_ascii=False), context))
+        return data
+
+    def render_extra_config(self, context):
+        """
+        Render extra_config and return a dict of Header: Value pairs.
+        Raises:
```
Comment (glennmatthews) [inline nautobot/extras/models/models.py:534]:
```suggestion
        Render extra_config and return a dict of Header: Value pairs.

        Raises:
```

File: nautobot/extras/models/models.py [hunk 20]
```diff
@@ -489,13 +489,66 @@ class ExternalIntegration(PrimaryModel):
...
+        """
+        if not self.extra_config:
+            return {}
+
+        data = json.loads(render_jinja2(json.dumps(self.extra_config, ensure_ascii=False), context))
+        return data
+
+    def render_remote_url(self, context):
+        """
+        Render remote_url in jinja2 templates.
```
Comment (glennmatthews) [inline nautobot/extras/models/models.py:547]:
```suggestion
        Render remote_url in jinja2 templates.

        Raises:
            TemplateAssertionError: Raised when an invalid template helper function exists in remote_url.
            TemplateSyntaxError: Raised when an invalid template variable exists in remote_url.
```

**Review by glennmatthews — COMMENTED** [review]
Please be sure to update `nautobot/docs/user-guide/platform-functionality/externalintegration.md` as well to document these new fields and APIs.

File: nautobot/users/factory.py [hunk 21]
```diff
@@ -13,27 +13,9 @@
...
-        ),
-        factory.Maybe(
-            "has_last_name",
-            factory.LazyAttribute(lambda u: u.last_name.lower()),
-            factory.Faker("user_name"),
-        ),
-    )
+        exclude = "has_email"
+
+    username = factory.Faker("user_name")
```
Comment (glennmatthews) [inline nautobot/users/factory.py:18]: Why this change? Removing the first_name/last_name properties from the generated user feels like a step backwards to me.

File: nautobot/extras/tests/test_models.py [hunk 22]
```diff
@@ -809,6 +810,50 @@ def test_timeout_validation(self):
...
+        ei = ei_with_extra_config.first()
+        ei.extra_config = {"context": "{% foo %}"}
+        ei.save()
+        with self.assertRaises(TemplateSyntaxError):
+            ei.render_extra_config({})
+
+        ei.extra_config = "{{ data | notvalid }}"
+        ei.save()
+        with self.assertRaises(TemplateAssertionError):
+            ei.render_extra_config({})
```
Comment (glennmatthews) [inline nautobot/extras/tests/test_models.py:833]: Am I missing something or do we not have any positive tests here (valid J2 template, data gets substituted as expected)?

**Review by glennmatthews — APPROVED** [review]

File: nautobot/extras/forms/forms.py [hunk 23]
```diff
@@ -605,10 +609,16 @@ class ExternalIntegrationBulkEditForm(NautobotBulkEditForm):
     verify_ssl = forms.NullBooleanField(required=False, label="Verify SSL", widget=BulkEditNullBooleanSelect)
     timeout = forms.IntegerField(required=False, min_value=0)
     extra_config = forms.JSONField(required=False)
+    http_method = forms.ChoiceField(
+        required=False,
+        label="HTTP Method",
+        choices=add_blank_choice(WebhookHttpMethodChoices),
+    )
+    headers = forms.JSONField(required=False, label="HTTP Request headers")
```
[the remark this answers is not in the document]
Comment (HanlinMiao) [reply nautobot/extras/forms/forms.py:617]: Sorry, I should have responded more promptly. My reasoning is that since we are using a custom template for Edit/Create ExternalIntegration, the form field's order are hard-coded in the template anyway. So I thought the order here on the form did not matter. Please correct me if I am wrong.

File: nautobot/users/factory.py (hunk 21 above)
Comment (HanlinMiao) [reply nautobot/users/factory.py:18]: Yeah, I think our old way of constructing a username was flaky and it was running into duplicate name issues in MySQL https://github.com/nautobot/nautobot/actions/runs/7149742147/job/19472323800. So I thought it might be more straightforward to just generate a user_name directly.

File: nautobot/extras/forms/forms.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (glennmatthews) [reply nautobot/extras/forms/forms.py:628]: Makes sense. Thanks :-)

File: nautobot/users/factory.py (hunk 21 above)
Comment (glennmatthews) [reply nautobot/users/factory.py:18]: Would be interesting to know what name the faker was running into issues with. I wonder if just changing it so that first_name and last_name were always present instead of randomly absent would be a cleaner fix. Not critical though.
